src/behavior-considered-undefined.md

```diff
@@ -23,44 +23,56 @@ code.
 
 * Data races.
-* Dereferencing a null or dangling raw pointer.
-* Unaligned pointer reading and writing outside of [`read_unaligned`]
-  and [`write_unaligned`].
-* Reads of [undef] \(uninitialized) memory.
-* Breaking the [pointer aliasing rules] on accesses through raw pointers;
-  a subset of the rules used by C.
-* `&mut T` and `&T` follow LLVM’s scoped [noalias] model, except if the `&T`
-  contains an [`UnsafeCell<U>`].
-* Mutating non-mutable data &mdash; that is, data reached through a shared
+* Dereferencing (using the `*` operator on) a dangling or unaligned raw pointer.
+* Breaking the [pointer aliasing rules]. `&mut T` and `&T` follow LLVM’s scoped
+  [noalias] model, except if the `&T` contains an [`UnsafeCell<U>`].
+* Mutating non-mutable data (that is, data reached through a shared
   reference or data owned by a `let` binding), unless that data is contained
   within an [`UnsafeCell<U>`].
-* Invoking undefined behavior via compiler intrinsics:
-  * Indexing outside of the bounds of an object with [`offset`] with
-    the exception of one byte past the end of the object.
-  * Using [`std::ptr::copy_nonoverlapping_memory`], a.k.a. the `memcpy32`and
-    `memcpy64` intrinsics, on overlapping buffers.
-* Invalid values in primitive types, even in private fields and locals:
-  * Dangling or null references and boxes.
+* Invoking undefined behavior via compiler intrinsics.
+* Executing code compiled with platform features that the current platform
+  does not support (see [`target_feature`]).
+* Unwinding into another language.
+* Producing an invalid value, even in private fields and locals. "Producing" a
+  value happens any time a value is assigned, passed to a function/primitive
+  operation or returned from a function/primitive operation.
+  The following values are invalid (at their respective type):
   * A value other than `false` (`0`) or `true` (`1`) in a `bool`.
   * A discriminant in an `enum` not included in the type definition.
+  * A null `fn` pointer.
   * A value in a `char` which is a surrogate or above `char::MAX`.
+  * A `!` (all values are invalid for this type).
+  * A dangling or unaligned reference or `Box`, or one that points to an invalid value.
+  * Invalid metadata in a wide reference, `Box` or raw pointer:
+    * slice metadata is invalid if the slice has a total size larger than
+      `isize::MAX` bytes in memory.
+    * `dyn Trait` metadata is invalid if it is not a pointer to a vtable for
+      `Trait` that matches the actual dynamic trait the reference points to.
   * Non-UTF-8 byte sequences in a `str`.
-* Executing code compiled with platform features that the current platform
-  does not support (see [`target_feature`]).
+  * [Uninitialized memory][undef] in the value of an integer (`i*`/`u*`),
+    floating point value (`f*`), or raw pointer.
+  * Invalid values for a type with a custom definition of invalid values, such
+    as a `NonNull` that is null. (Requesting custom invalid values is an
```

Comment:
I don't understand what you are suggesting I do here. Sure the rustc guide should explain that if it does not already, but this also must be called out here.

Reply:
I'm suggesting that "(Requesting custom...)" should be moved to a note or removed as that ability is unstable. If you have a note then it should point to the relevant rustc guide section.

Reply:
What do you mean by "note"? It already is a parenthetical remark.

I think this is needed here because otherwise people reading this will be confused. Most users of Rust do not and should not know about this feature.

Maybe @gnzlbg was right and we should just also exhaustively list the affected types here. @Gankra didn't like this but such stylistic differences between Nomicon and Reference seem fine.

Reply:
> What do you mean by "note"? It already is a parenthetical remark.

https://doc.rust-lang.org/nightly/reference/#conventions

How rustc achieves the fact that certain types have various niches is not part of the language definition.

Reply:
This is a document for human beings, not robots.

Reply:
This parenthetical also does not fit "useful information about the state of the book or [...] useful, but mostly out of scope, information".

Reply:
> This is a document for human beings, not robots.

I don't know what you mean by this.

Implementation details of rustc belong in notes; that's the convention in the reference that is used. cc @Havvy

Reply:
I mean that this should be readable, so we have to mention `NonNull` here and not send the user on a chase for looking up what "custom definition of invalid values" means.

Reply:
I don't understand how a note in the conventional style I linked above about how rustc achieves this would be unreadable (it's also not necessary to say it is unstable, only a note re. the names of the attributes + links with elaboration in the rustc guide).

In other words I think this should be changed into:

> **Note:** In `rustc`, such custom definitions are achieved with [`rustc_layout_scalar_valid_range_start`](link_to_rustc_guide), other thing, ..., and last thing.

Reply:
I don't think the attribute is very interesting to most users; the types are. Also if this is what notes are for, I think the def.n you linked me to need adjustment; this is not "mostly out of scope". It is relevant for the vast majority of readers.

Is there an example for how a note rendered in a nested enumeration looks like?

Reply:
> I don't think the attribute is very interesting to most users; the types are.

But you've already mentioned one of those types. Saying that this is achieved with some unstable attribute without mentioning the name of that attribute is *vague* and doesn't say what the other types are. I think it's far easier to know the name of the attribute and then search for it in the source if you want to know what the full list of types are. Alternatively, mention all of the base types.

> Is there an example for how a note rendered in a nested enumeration looks like?

See https://github.com/rust-lang-nursery/reference/search?p=2&q=note&type=&utf8=%E2%9C%93 for a list of notes. I didn't see any example in a nested enumeration when skimming.

Reply:
Something like this?

Reply:
Yeah that's better.